runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -2937,19 +2937,264 @@ J9::Z::TreeEvaluator::anewArrayEvaluator(TR::Node * node, TR::CodeGenerator * cg
+   bool use64BitClasses = comp->target().is64Bit() && !TR::Compiler->om.generateCompressedObjectHeaders();
+
+   // Init class and fields, then jump to end of ICF
+   generateRXInstruction(cg, use64BitClasses ? TR::InstOpCode::STG : TR::InstOpCode::ST, node, classReg, generateS390MemoryReference(targetReg, TR::Compiler->om.offsetOfObjectVftField(), cg));
+   generateSILInstruction(cg, use64BitClasses ? TR::InstOpCode::MVGHI : TR::InstOpCode::MVHI, node, generateS390MemoryReference(targetReg, fej9->getOffsetOfContiguousArraySizeField(), cg), 0);
+   generateSILInstruction(cg, use64BitClasses ? TR::InstOpCode::MVGHI : TR::InstOpCode::MVHI, node, generateS390MemoryReference(targetReg, fej9->getOffsetOfDiscontiguousArraySizeField(), cg), 0);
+   generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BRC, node, cFlowRegionEnd);
+
+   // We end up in this region of the ICF if the first dimension is non-zero and the second dimension is zero.
+   generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, nonZeroFirstDimLabel);
+
+   TR::Register *componentClassReg = cg->allocateRegister();
+   generateRXInstruction(cg, TR::InstOpCode::getLoadOpCode(), node, componentClassReg, generateS390MemoryReference(classReg, offsetof(J9ArrayClass, componentType), cg));
+
+   // Calculate maximum allowable object size in elements and ump to OOL if it's higher than firstDimLenReg.
+   int32_t elementSize = comp->useCompressedPointers() ? TR::Compiler->om.sizeofReferenceField() : static_cast<int32_t>(TR::Compiler->om.sizeofReferenceAddress());
+   uintptr_t maxObjectSize = cg->getMaxObjectSizeGuaranteedNotToOverflow();
+   uintptr_t maxObjectSizeInElements = maxObjectSize / elementSize;
+
+   if (comp->target().is64Bit() && !(maxObjectSizeInElements >= 0 && maxObjectSizeInElements <= (uintptr_t)INT_MAX))
+      {
+      uint32_t high32 = maxObjectSizeInElements >> 32;
+      generateRILInstruction(cg, TR::InstOpCode::LLIHF, node, temp1Reg, high32);
+      generateRILInstruction(cg, TR::InstOpCode::IILF, node, temp1Reg, static_cast<uint32_t>(maxObjectSizeInElements));
+      generateRIEInstruction(cg, TR::InstOpCode::CLGRJ, node, firstDimLenReg, temp1Reg, oolJumpLabel, TR::InstOpCode::COND_BRH);
+      }
+   else
+      {
+      generateRILInstruction(cg, TR::InstOpCode::CFI, node, firstDimLenReg, static_cast<uint32_t>(maxObjectSizeInElements));
+      generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BHR, node, oolJumpLabel);
+      }
+
+   // Now check to see if we have enough space to do the allocation. If not then jump to OOL code.
+   generateRRInstruction(cg, TR::InstOpCode::LR, node, temp1Reg, firstDimLenReg);
+
+   int32_t round = (elementSize >= TR::Compiler->om.objectAlignmentInBytes()) ? 0 : TR::Compiler->om.objectAlignmentInBytes();
+   int32_t disp32 = round ? (round - 1) : 0;
+
+   uint8_t _multiplierToStrideMap[] = {0, 0, 1, 0, 2, 0, 0, 0, 3};
+   uint8_t shiftVal = _multiplierToStrideMap[elementSize];
```

Comment:
We should inline this variable as it is only used once.

Reply:
Done in: 0d2a7a6